tests/k8st/test_base.py

```diff
@@ -0,0 +1,182 @@
+        return client.CoreV1Api()
+
+    def check_pod_status(self, ns):
+        pods = self.cluster.list_namespaced_pod(ns)
+
+        for pod in pods.items:
+            logger.info("%s\t%s\t%s", pod.metadata.name, pod.metadata.namespace, pod.status.phase)
+            assert pod.status.phase == 'Running'
+
+    def create_namespace(self, ns_name):
+        self.cluster.create_namespace(client.V1Namespace(metadata=client.V1ObjectMeta(name=ns_name)))
+
+    def create_service(self, image, name, ns, port, replicas=2):
+        cluster = self.k8s_client()
+        cluster.create_namespace(client.V1Namespace(metadata=client.V1ObjectMeta(name=ns)))
+
+        # Run a deployment with <replicas> copies of <image>, with the
+        # pods labelled with "app": <name>.
+        deployment = client.ExtensionsV1beta1Deployment(
+            api_version="extensions/v1beta1",
+            kind="Deployment",
+            metadata=client.V1ObjectMeta(name=name),
+            spec=client.ExtensionsV1beta1DeploymentSpec(
+                replicas=replicas,
+                template=client.V1PodTemplateSpec(
+                    metadata=client.V1ObjectMeta(labels={"app": name}),
+                    spec=client.V1PodSpec(containers=[
+                        client.V1Container(name=name,
+                                           image=image,
+                                           ports=[client.V1ContainerPort(container_port=port)]),
+                    ]))))
+        api_response = client.ExtensionsV1beta1Api().create_namespaced_deployment(
+            body=deployment,
+            namespace=ns)
+        logger.debug("Deployment created. status='%s'" % str(api_response.status))
+
+        # Create a service called <name> whose endpoints are the pods
+        # with "app": <name>; i.e. those just created above.
+        service = client.V1Service(
+            metadata=client.V1ObjectMeta(
+                name=name,
+                labels={"name": name},
+            ),
+            spec={
+                "ports": [{"port": port}],
+                "selector": {"app": name},
+            }
+        )
+        api_response = cluster.create_namespaced_service(
+            body=service,
+            namespace=ns,
+        )
+        logger.debug("Service created. status='%s'" % str(api_response.status))
+
+    @classmethod
+    def check_calico_version(cls):
+        config.load_kube_config(os.environ.get('KUBECONFIG'))
+        api = client.AppsV1Api(client.ApiClient())
+        node_ds = api.read_namespaced_daemon_set("calico-node", "kube-system", exact=True, export=True)
+        for container in node_ds.spec.template.spec.containers:
+            if container.name == "calico-node":
+                if container.image != "calico/node:latest-amd64":
+                    container.image = "calico/node:latest-amd64"
+                    api.replace_namespaced_daemon_set("calico-node", "kube-system", node_ds)
+                    time.sleep(3)
```

Comment:
Is this a stray sleep? We do some retries below which should handle this?

Reply:
I guess the idea is to ensure that the replacement has _started_.  Would that make sense?